### ISSUE TITLE
Avoid new user loss-of-work by reminding project save isn't layer save

### ISSUE THREAD

**Issue by Houska1** [qgis/QGIS#41535]
QGIS distinction between project and layer, and distinct saving mechanisms, is great. And it's unsurprising for those of us with database backgrounds (front end vs data store) -- or who actually read the manual.

But there are several recent reports of new users who lose a lot of effort and get frustrated by the following failure mode:
- opening layers for editing
- saving the project but not the layers
- leaving QGIS open for hours/days until it crashes. 

They think their project saves have saved "everything" and -- if they don't close down QGIS for days -- don't get the usual warning on orderly project close/exit. _Examples:_ [today on reddit](https://www.reddit.com/r/QGIS/comments/lih6rr/saving_policy/) and [a few months back on gis.se](https://gis.stackexchange.com/questions/375419/qgis-saved-project-crashed-and-i-lost-three-days-of-work/375429)

A fairly simple educational reminder would be:

- A userprofile-level advanced option called something like `UI/warnLayerProjectUnsaved` defaulting `true`.
- If user saves project, and has any unsaved layer edits at that time AND this advanced option is `true`, a dialog comes up warning them to also save layer edits, and giving them the option to "Don't warn me again" which would set this advanced option to `false`

Bottom line, let's avoid new users starting with a harrowing experience. And let's recognize many new users might come to us with an office software vs database software background -- and keeping your Word (or whatever document) open for days while just hitting the "save" icon anytime you do something is pretty commonplace.

**Comment by roya0045** [comment]
I feel like instilling the habit of using ctrl+s is important. When you work on remote data and your connection can drop at anytime or qgis can crash you hit ctrl+s when you do any significant work. It's a thing you should learn at least once. 

If the usual dialog doesn't appear when you quit the project, it might have been an edge case. I always got the 'you have unsaved edits' dialog when closing.

The next step would be going with autosave and I don't think that's a good idea as sometimes you want to lose your edits.

**Comment by Houska1** [comment]
Agree that you sometimes do want to lose your edits, so "automatic autosave" would not be a good solution.

However, to get the "you have unsaved edits" warning you need to quit QGIS or at least close the project. 

A frequent usage pattern for desktop office software is "keep open indefinitely and just hit save" - I'm not sure I ever closed Microsoft Word this week, and had the report I'm working on open all or most of the time. I think the same usage pattern is more frequent for QGIS than old hands realize --- and let's face it, QGIS does crash more often than it should (which is when this confusion leads to actual data loss).

That's why I feel we should be more new user friendly about this, especially when the negative impact on the more experienced is one dialog box to answer once during the lifetime of their userprofile.

**Comment by uclaros** [comment]
>     * If user saves project, and has any unsaved layer edits at that time AND this advanced option is `true`, a dialog comes up warning them to also save layer edits, and giving them the option to "Don't warn me again" which would set this advanced option to `false`

That's a nice idea! Helpful for the newcomers and minimally obtrusive to the experienced.

**Comment by roya0045** [comment]
The safest option would be to cache unsaved edits in a project. On discard save you clear the cache. Otherwise on crash when you reopen the project you prompt for unsaved edits and reload them or discard them.

This would be the safest approach.

**Comment by joolsmtw** [comment]
> 
> 
> I feel like instilling the habit of using ctrl+s is important. When you work on remote data and your connection can drop at anytime or qgis can crash you hit ctrl+s when you do any significant work. It's a thing you should learn at least once.
> 
> If the usual dialog doesn't appear when you quit the project, it might have been an edge case. I always got the 'you have unsaved edits' dialog when closing.
> 
> The next step would be going with autosave and I don't think that's a good idea as sometimes you want to lose your edits.

As one of the people mentioned that the original problem happened to: I've had my habit of CTRL+S for a long time, but CTRL+S only saves the project, not the individual layers in QGIS, does it not? At least that has been the original problem for me, as a learning-by-doing kind of person and not coming from a database management layout (where saving individual layers seems to be commonplace, or at least a comparable situation, probably referring to comitting individual changes etc.). I simply forgot to properly save my individual layers, trusting that CTRL+S did the job already.

A warning to be aware of this situation, as the original poster suggested, seems pretty sweet and easy to implement to me. :)

**Comment by roya0045** [comment]
@joolsmtw Isn't the red pen/yellow pen symbol on layer in edit mode sufficient to show that you have unsaved edits?

**Comment by joolsmtw** [comment]
> 
> 
> @joolsmtw Isn't the red pen/yellow pen symbol on layer in edit mode sufficient to show that you have unsaved edits?

Not for absolute greenies – by now it is, after sufficient study of documentation.

**Comment by TurboGraphxBeige** [comment]
> > @joolsmtw Isn't the red pen/yellow pen symbol on layer in edit mode sufficient to show that you have unsaved edits?
> 
> Not for absolute greenies – by now it is, after sufficient study of documentation.

Indeed, took me a little while to notice the different colors in the pen icon.

> If user saves project, and has any unsaved layer edits at that time AND this advanced option is true, a dialog comes up warning them to also save layer edits, and giving them the option to "Don't warn me again" which would set this advanced option to false

I also think it's a good idea. It's a friendly reminder for newcomers and absolutely not annoying for experienced users as it happens only once. It also seems quite easy to implement. 

Why not!

**Comment by roya0045** [comment]
I guess then pen thing is only evocative for people that used arcgis previously as the behaviour is quite similar.

**Comment by realpixelcode** [comment]
> sometimes you want to lose your edits

There should be an explicit button to undo/lose undesired actions/edits. Relying on QGIS' inability to auto-save unsaved work is just a hacky **workaround**, not a feature to be proud of.

---

I just had the following issue in v3.44.7:

1. In an existing project, I added a new point layer.
2. I added some points to the layer and hit the “save layer edits” button.
3. I saved the project and closed QGIS.
4. The next time I opened QGIS, the layer was present but without any points in it.

Is this because I didn't explicitly persist/export the layer? Is it the same issue we're discussing here?

**Comment by agiudiceandrea** [comment]
@realpixelcode , which kind of layer did add to the project? Was it a "New **Temporary** Scratch Layer"? If yes, have you read the warning ("Warning: Temporary scratch layer are not saved and will be discarded when QGIS is closed") in the layer creation dialog window? Have you read the warning ("This project includes one or more temporary scratch layers. These layers are not saved to disk and their contents will be permanently lost. Are you sure you want to proceed?") displayed when a project containing Temporary Scratch layers is closed?

**Comment by realpixelcode** [comment]
It was just a regular vector layer, not a temporary one. There was no warning about anything.

**Comment by agiudiceandrea** [comment]
@realpixelcode , which format exactly was the vector layer? Please provide an example QGIS project and layer and step-by-step instruction to reproduce the reported issue.
